ecosystem/indexer-grpc/indexer-grpc-utils/src/in_memory_cache.rs

```diff
@@ -0,0 +1,439 @@
+                    },
+                    _ = tokio::time::sleep(std::time::Duration::from_millis(IN_MEMORY_CACHE_LOOKUP_RETRY_INTERVAL_MS)) => {
+                        continue;
+                    },
+                }
+            }
+            let redis_waiting_duration = current_time.elapsed().as_secs_f64();
+            let start_time = std::time::Instant::now();
+            let end_version = std::cmp::min(
+                current_latest_version,
+                in_cache_latest_version + 10 * MAX_REDIS_FETCH_BATCH_SIZE as u64,
+            );
+            let versions_to_fetch = (in_cache_latest_version..end_version).collect();
+            let transactions = batch_get_transactions(&mut conn, versions_to_fetch, storage_format)
+                .await
+                .unwrap();
+            // Ensure that transactions are ordered by version.
+            let cache_processing_start_time = std::time::Instant::now();
+            let mut newly_added_bytes = 0;
+            for (ind, transaction) in transactions.iter().enumerate() {
+                if transaction.version != in_cache_latest_version + ind as u64 {
+                    panic!("Transactions are not ordered by version");
+                }
+                newly_added_bytes += transaction.encoded_len() as u64;
+            }
+
+            for transaction in transactions {
+                cache.insert(transaction.version, Arc::new(transaction));
+            }
+            let processing_duration = start_time.elapsed().as_secs_f64();
+            tracing::info!(
+                redis_latest_version = current_latest_version,
+                in_memory_latest_version = in_cache_latest_version,
+                new_in_memory_latest_version = end_version,
+                processing_duration,
+                cache_processing_duration = cache_processing_start_time.elapsed().as_secs_f64(),
+                redis_waiting_duration,
+                "In-memory cache is updated"
+            );
+            let start_time = std::time::Instant::now();
+            let mut current_cache_metadata = { *cache_metadata.read().await };
+            current_cache_metadata.latest_version = end_version;
+            current_cache_metadata.total_size_in_bytes += newly_added_bytes;
+            // Get the data available.
+            *cache_metadata.write().await = current_cache_metadata;
+            // Clean up.
+            while current_cache_metadata.total_size_in_bytes > MAX_IN_MEMORY_CACHE_CAPACITY_IN_BYTES
+            {
+                let key_to_remove = current_cache_metadata.first_version;
+                let (_k, v) = cache
+                    .remove(&key_to_remove)
+                    .expect("Failed to remove the key");
+                current_cache_metadata.total_size_in_bytes -= v.encoded_len() as u64;
+                current_cache_metadata.first_version += 1;
+            }
+            *cache_metadata.write().await = current_cache_metadata;
```

Comment:
Should we update the metadata first?

Reply:
the update happens twice: `last_version` only first and then after the garbage collection is done. 

for readers, if the version trying to access is evicted, fallback to other storages.